pkgs/stdenv/generic/setup.sh

```diff
@@ -400,8 +400,14 @@ concatTo() {
                 -a*)
                     targetref+=( "${nameref[@]}" ) ;;
                 *)
-                    # shellcheck disable=SC2206
-                    targetref+=( ${nameref-} ) ;;
+                    if [[ "$name" = *"Array" ]]; then
+                        nixErrorLog "concatTo(): $name is not declared as array, treating as a singleton. This will become an error in future"
+                        targetref+=( "${nameref-}" )
+                    else
+                        # shellcheck disable=SC2206
+                        targetref+=( ${nameref-} )
+                    fi
+                    ;;
```

Comment:
@emilazy @tie ok to me this is super innocuous I mean what can go wrong testing string for a suffix, but tell me did I step on any of the bash landmines

Reply:
The Bash looks fine to me. Is there any chance we could make the warning happen at evaluation time so it’s more visible?

Reply:
Huh. I guess we could add a warning in `mkDerivation`, and that'd also make Ofborg light up all (ab)users in Nixpkgs. The fallback, if we choose to go with the fallback and we probably do because of out-of-tree users, is still easiest to implement in bash though

Reply:
Yeah, the fallback is good. We should probably even warn in Bash too, because you could set the wrong thing in bash code. But a `mkDerivation` warning would be nice if it’s not too invasive.

Reply:
I think this should be using `${arr[@]}` notation for `*Array` variables.

Reply:
The `sudo` had an issue because it set `*Array` variable as a derivation input, so it got exported as a string and with `${var[@]}` expansion strings are treated as single-element arrays.

Reply:
>  Is there any chance we could make the warning happen at evaluation time so it’s more visible?

Never touched `check-meta.nix` before, opening a separate PR targetting master for this discussion: https://github.com/NixOS/nixpkgs/pull/335110

> I think this should be using ${arr[@]} notation for *Array variables.

Cannot think of a single case test case where this would produce a different result, but I recovered the original code (including the `${nameref+...}` part) verbatim

Reply:
Sorry I didn't reply before the merge, I was away. I agree to some extend, and I did start with treating the input as just a string. But then I do hope to remove this branch maybe after a release, and the intent of reintroducing this branch was to mimic exactly what the old implementation was doing so... I guess it's OK we merged it the way it was